@here/harp-geometry/lib/ClipLineString.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Math2D } from "@here/harp-utils";
+import { Vector2 } from "three";
+
+/**
+ * A clipping edge.
+ *
+ * @remarks
+ * Clip lines using the Sutherland-Hodgman algorithm.
+ *
+ * @internal
+ */
+class ClipEdge {
+    readonly p0: Vector2;
+    readonly p1: Vector2;
+
+    /**
+     * Creates a clipping edge.
+     *
+     * @param x1 - The x coordinate of the first point of this ClipEdge.
+     * @param y1 - The y coordinate of the first point of this ClipEdge.
+     * @param x2 - The x coordinate of the second point of this ClipEdge.
+     * @param y2 - The y coordinate of the second point of this ClipEdge.
+     * @param isInside - The function used to test points against this ClipEdge.
+     */
+    constructor(
+        x1: number,
+        y1: number,
+        x2: number,
+        y2: number,
+        private readonly isInside: (p: Vector2) => boolean
+    ) {
+        this.p0 = new Vector2(x1, y1);
+        this.p1 = new Vector2(x2, y2);
+    }
+
+    /**
+     * Tests if the given point is inside this clipping edge.
+     */
+    inside(point: Vector2): boolean {
+        return this.isInside(point);
+    }
+
+    /**
+     * Computes the intersection of a line and this clipping edge.
+     *
+     * @remarks
+     * {@link https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection
+     *    | line-line intersection}.
+     */
+    computeIntersection(a: Vector2, b: Vector2): Vector2 {
+        const result = new Vector2();
+        Math2D.intersectLines(
+            a.x,
+            a.y,
+            b.x,
+            b.y,
+            this.p0.x,
+            this.p0.y,
+            this.p1.x,
+            this.p1.y,
+            result
+        );
+        return result;
+    }
+
+    /**
+     * Clip the input line against this edge.
+     */
+    clipLine(lineString: Vector2[]): Vector2[][] {
+        const inputList = lineString;
+
+        const result: Vector2[][] = [];
+
+        lineString = [];
+        result.push(lineString);
+
+        const pushPoint = (point: Vector2) => {
+            if (lineString.length === 0 || !lineString[lineString.length - 1].equals(point)) {
+                lineString.push(point);
+            }
+        };
+
+        for (let i = 0; i < inputList.length; ++i) {
+            const currentPoint = inputList[i];
+            const prevPoint = i > 0 ? inputList[i - 1] : undefined;
+
+            if (this.inside(currentPoint)) {
+                if (prevPoint !== undefined && !this.inside(prevPoint)) {
+                    if (lineString.length > 0) {
+                        lineString = [];
+                        result.push(lineString);
+                    }
+                    pushPoint(this.computeIntersection(prevPoint, currentPoint));
```

Comment:
`computeIntersection` can returned undefined, should this be checked?

Reply:
it should not be needed because `computeIntersection` is called only after the end points are tested against the clipping edge and an intersection is known to exist.